dsl/flatten.cc

```diff
@@ -0,0 +1,340 @@
+                    exprs[i - 1].staticLevel = expr.staticLevel;
+                }
+            }
+        }
+
+        // these will store the bodies of the `class << self` blocks we create at the end
+        vector<ast::ClassDef::RHS_store> nestedBlocks;
+        for (int level = 2; level <= highestLevel; level++) {
+            nestedBlocks.emplace_back();
+        }
+
+        // this vector contains all the possible RHS target locations that we might move to
+        vector<ast::ClassDef::RHS_store *> targets;
+        // 0 and 1 both go into the class itself
+        targets.emplace_back(&rhs);
+        targets.emplace_back(&rhs);
+        // 2 and up go into the to-be-created `class << self` blocks
+        for (auto &tgt : nestedBlocks) {
+            targets.emplace_back(&tgt);
+        }
+
+        // move everything to its appropriate target
+        for (auto &expr : exprs) {
+            if (auto methodDef = ast::cast_tree<ast::MethodDef>(expr.expr.get())) {
+                methodDef->setIsSelf(expr.staticLevel > 0);
+            }
+            targets[expr.staticLevel]->emplace_back(std::move(expr.expr));
+        }
+
+        // generate the nested `class << self` blocks as needed and add them to the class
+        for (auto &body : nestedBlocks) {
+            auto classDef =
+                ast::MK::Class(loc, loc,
+                               make_unique<ast::UnresolvedIdent>(core::Loc::none(), ast::UnresolvedIdent::Class,
+                                                                 core::Names::singleton()),
+                               {}, std::move(body), ast::ClassDefKind::Class);
+            rhs.emplace_back(std::move(classDef));
+        }
+
+        return rhs;
+    }
+
+    int computeStaticLevel(ast::MethodDef *methodDef) {
+        auto &methods = curMethodSet();
+        int prevLevel = methods.stack.size() > 0 ? methods.stack.back().staticLevel : 0;
+        return prevLevel + (methodDef->isSelf() ? 1 : 0);
+    }
+
+    struct MethodData {
+        // This is an index into the methods stack
+        int idx;
+        // we need to keep information around about whether we're in a static outer context: for example, if we have
+        //
+        //   def self.foo; def bar; end; end
+        //
+        // then we should flatten it to
+        //
+        //  def self.foo; end
+        //  def self.bar; end
+        //
+        // which means when we get to `bar` we need to know that the outer context `foo` is static. We pass that down
+        // the current stack by means of this `isStatic` variable.
+        int staticLevel;
+        MethodData(int idx, int staticLevel) : idx(idx), staticLevel(staticLevel){};
+    };
+    struct MovedItem {
+        unique_ptr<ast::Expression> expr;
+        int staticLevel;
+        MovedItem(unique_ptr<ast::Expression> expr, int staticLevel) : expr(move(expr)), staticLevel(staticLevel){};
+        MovedItem() = default;
+    };
```

Comment:
This `MovedItem` struct combines both an expression to be moved with a 'staticness level' to move it to (0 being an instance method, 1 being a static method, 2 being a static method on the metaclass, etc.) but we won't actually need this metadata when we switch to the correct semantics, and this whole struct can just get replaced with a `unique_ptr<ast::Expression>` everywhere it gets used.